src/gpio.rs

```diff
@@ -50,1205 +46,1289 @@ pub trait GpioExt {
+    }
+
+    pub trait Otyper {
+        fn push_pull(&mut self, i: u8);
+        fn open_drain(&mut self, i: u8);
+    }
+
+    pub trait Ospeedr {
+        fn low(&mut self, i: u8);
+        fn medium(&mut self, i: u8);
+        fn high(&mut self, i: u8);
+    }
+
+    pub trait Pupdr {
+        fn floating(&mut self, i: u8);
+        fn pull_up(&mut self, i: u8);
+        fn pull_down(&mut self, i: u8);
+    }
+
+    pub trait Afr {
+        fn afx(&mut self, i: u8, x: u8);
+    }
+
+    pub trait Gpio {
+        type Reg: GpioRegExt + ?Sized;
+
+        fn ptr(&self) -> *const Self::Reg;
+    }
 }
 
-/// Floating input (type state)
-pub struct Floating;
-/// Pulled down input (type state)
-pub struct PullDown;
-/// Pulled up input (type state)
-pub struct PullUp;
-
-/// Output mode (type state)
-pub struct Output<MODE> {
-    _mode: PhantomData<MODE>,
+use private::{Afr, GpioRegExt, Moder, Ospeedr, Otyper, Pupdr};
+
+/// Marker trait for GPIO ports
+pub trait Gpio: private::Gpio {}
+
+/// Marker trait for compile time defined GPIO ports
+pub trait GpioStatic: Gpio {
+    /// Associated MODER register
+    type MODER: Moder;
+    /// Associated OTYPER register
+    type OTYPER: Otyper;
+    /// Associated OSPEEDR register
+    type OSPEEDR: Ospeedr;
+    /// Associated PUPDR register
+    type PUPDR: Pupdr;
 }
 
-/// Push pull output (type state)
-pub struct PushPull;
-/// Open drain output (type state)
-pub struct OpenDrain;
+/// Marker trait for pin number
+pub trait Index {
+    #[doc(hidden)]
+    fn index(&self) -> u8;
+}
 
-/// Analog mode (type state)
-pub struct Analog;
+impl<U> Index for U
+where
+    U: Unsigned,
+{
+    #[inline(always)]
+    fn index(&self) -> u8 {
+        Self::U8
+    }
+}
 
-/// Alternate function 0 (type state)
-pub struct AF0;
+/// Marker trait for readable pin modes
+pub trait Readable {}
 
-/// Alternate function 1 (type state)
-pub struct AF1;
+/// Marker trait for slew rate configurable pin modes
+pub trait OutputSpeed {}
 
-/// Alternate function 2 (type state)
-pub struct AF2;
+/// Marker trait for internal resistor enabled pin modes
+pub trait InternalResistor {}
 
-/// Alternate function 3 (type state)
-pub struct AF3;
+/// Runtime defined GPIO port (type state)
+pub struct Gpiox {
+    ptr: *const dyn GpioRegExt,
+}
 
-/// Alternate function 4 (type state)
-pub struct AF4;
+impl private::Gpio for Gpiox {
+    type Reg = dyn GpioRegExt;
 
-/// Alternate function 5 (type state)
-pub struct AF5;
+    fn ptr(&self) -> *const Self::Reg {
+        self.ptr
+    }
+}
 
-/// Alternate function 6 (type state)
-pub struct AF6;
+impl Gpio for Gpiox {}
 
-/// Alternate function 7 (type state)
-pub struct AF7;
+/// Runtime defined pin number (type state)
+pub struct Ux(u8);
 
-/// Alternate function 8 (type state)
-pub struct AF8;
+impl Index for Ux {
+    fn index(&self) -> u8 {
+        self.0
+    }
+}
 
-/// Alternate function 9 (type state)
-pub struct AF9;
+/// Input mode (type state)
+pub struct Input;
```

Comment:
Maybe add a documentation notice, that the Input type is not encoded in a type state anymore?
The user has to know, that it is their responsibility to set the correct input type and that this 
is due issues explained in https://github.com/stm32-rs/stm32f3xx-hal/issues/125. 

The change notice belonging to the changelog, the overall behavior should be written down here.

Reply:
Yeah, documentation update is needed.

However, I currently think about re-adding `into_input_xxx()` methods which implemented like
```rust
pub fn into_input_floating(self, moder: &mut GPIO::MODER, pupdr: &mut GPIO::PUPDR) -> Pin<GPIO, INDEX, Input> {
    pupdr.floating(self.index.index());
    self.into_input(moder)
}
```